### DISCUSSION TITLE
Add build number?

### DISCUSSION THREAD

**Discussion by MSP-Greg** [community/community#26689] [Actions]
<p>First, just because other CI providers have a feature doesn’t necessarily mean that it should be implemented in Actions.</p>
<p>That said, it would be very helpful to have a build number, or maybe even a build.job number, similar to Travis.</p>
<p>Recently I wanted to look over the results of several builds, and there is no easy way to determine what is what (or what the order was).  If the ‘number’ appeared early in the html title, one could see it on the browser tab, which is very helpful…</p>

**Comment by MChevy422** [comment]
<p>Hello <span><a class="mention" href="/msp-greg">@msp-greg</a></span>, <br><br>Thank you for your feedback! We’re always working to improve GitHub and the GitHub Community Forum, and we consider every suggestion we receive. I’ve logged your feedback in our internal feature request list. </p>
<p>Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.<br><br>Once again, thank you for your input!<br><br>Greatly appreciated,<br><br>MChevy422</p>

**Comment by xt0rted** [comment]
<p>I’ve been using the build number as part of my version numbers (<em>yyyy.m.d.build</em>) which makes it easy to know when the deployment happened, and which build it’s from. This is one of the last things stopping me from moving all of my builds over to actions.</p>

**Comment by warrenbuckley** [comment]
<p>I have posted the same thing here over on the GitHub actions/toolkit repo as an issue as I am stil unsure where the best place is to discuss with the GitHub team and give feedback.</p>
<p>Is it on issuses or here in this forum?</p>
<p><a href="https://github.com/actions/toolkit/issues/65" rel="nofollow noopener">https://github.com/actions/toolkit/issues/65</a></p>

**Comment by AndreaGriffiths11** [comment]
<p>Thanks <a class="mention" href="/xt0rted">@xt0rted</a> and <a class="mention" href="/warrenbuckley">@warrenbuckley</a> for being here! This is the right place for this type of feedback as the team who directly works in actions is here and interacts with users. I can tell you that although as you know this is not available as yet, there is an internal discussion to add this feature to GitHub Actions and your feedback has been added to it.</p>

**Comment by ArturT** [comment]
<p>CI build ID would be helpful. I recently added integration for GitHub Actions into my ruby gem knapsack_pro for automated tests split across parallel jobs. The CI build ID is needed to allow to run a few CI build at the same time for the same git commit. </p>
<p>Here is more about my integration with GitHub Actions</p>
<p><a href="https://docs.knapsackpro.com/2019/how-to-run-rspec-on-github-actions-for-ruby-on-rails-app-using-parallel-jobs" rel="nofollow noopener">https://docs.knapsackpro.com/2019/how-to-run-rspec-on-github-actions-for-ruby-on-rails-app-using-parallel-jobs</a></p>

**Comment by einaregilsson** [comment]
<p>I’ve made a simple action to generate sequential build numbers, <strong><a href="https://github.com/einaregilsson/build-number" rel="nofollow noopener">https://github.com/einaregilsson/build-number</a> .</strong> Would love for it to be made obsolete by GitHub implementing this natively, but until then I use it so I can continue the same versioning scheme I used on my old CI/CD system. I don’t think there’s any way right now to make the build number appear in the list of workflow runs though, that would be great if that was possible.</p>

**Comment by ivailop** [comment]
<p>I think having <code>check-suite-id</code> (and <code>check-run-id</code>) available as environment variables would provide approptiate for the purpose number. </p>
<p>See <a href="https://github.community/t/check-suite-identifier-as-an-environment-variable/16642">https://github.community/t5/GitHub-Actions/Check-Suite-identifier-as-an-environment-variable/m-p/35498#M2240</a></p>

**Comment by sarunw** [comment]
<p>Looking forward for this implementation.</p>

**Comment by dfdgsdfg** [comment]
<p>Here is my workaround using pushed time for build number.</p>
<pre><code>- name: set build number
        run: |
          mkdir ../build
          echo $(( ($PUSHED_TIME - 1572572000) / 10 )) &gt; ../build/FULTTER_VERSION_CODE_BY_PUSHED_TIME
          echo "flutter.versionCode=$(cat ../build/FULTTER_VERSION_CODE_BY_PUSHED_TIME)" &gt;&gt; local.properties
        env:
          PUSHED_TIME: ${{ toJson(github.event.repository.pushed_at) }}
</code></pre>
<p><a href="https://medium.com/@esplo/using-an-unusual-timestamp-as-an-auto-generated-version-code-for-android-apps-f231b026edbd" rel="nofollow noopener">https://medium.com/@esplo/using-an-unusual-timestamp-as-an-auto-generated-version-code-for-android-apps-f231b026edbd</a></p>

**Comment by blake-newman** [comment]
<p>Here is our solution to this issue:</p>
<p><a href="https://medium.com/attest-engineering/adding-a-unique-github-build-identifier-7aa2e83cadca" rel="nofollow noopener">https://medium.com/attest-engineering/adding-a-unique-github-build-identifier-7aa2e83cadca</a></p>

**Comment by xt0rted** [comment]
<p>On the <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener">github context</a> there’s now a <strong>run_id</strong> and <strong>run_number</strong>  property that’s available to be passed into your actions. I’m not sure if you can get to this from one of the environment variables, but it’s available through the context at least.</p>
<p>There doesn’t seem to be docs on this yet but you can see it if you dump the context out like so:</p>
<pre><code>- name: Dump GitHub context
  env:
    GITHUB_CONTEXT: ${{ toJson(github) }}
  run: echo "$GITHUB_CONTEXT"
</code></pre>
<p><strong>run_id</strong> is the unique id of each run</p>
<p><strong>run_number</strong> looks to be an auto incrementing value specific to each workflow. This value doesn’t increment if you rerun the workflow.</p>

**Comment by peteraritchie** [comment]
<p>Is this not what $GITHUB_RUN_NUMBER is for?</p>

**Comment by liorhasson** [comment]
<p>Great!<br>
Also it wouldn’t be a full solution without adding the <strong>name</strong> and <strong>number</strong> in the UI under the Actions tab. Today only the workflow name exists in each row.</p>
<p>exp.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/15028bf8eb0f3b4326ea24a1f8486d20eeae3758.png" data-download-href="/uploads/short-url/2ZRtXOvauzdQXidlR8SzNE2l18I.png?dl=1" title="Screen Shot 2020-01-29 at 8.53.55.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/492604/181099238-8d7c680e-3be7-492d-8be5-54cd9f22217e.png" alt="Screen Shot 2020-01-29 at 8.53.55.png" data-base62-sha1="2ZRtXOvauzdQXidlR8SzNE2l18I" width="690" height="67" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/1/15028bf8eb0f3b4326ea24a1f8486d20eeae3758_2_690x67.png, https://aws1.discourse-cdn.com/github/optimized/2X/1/15028bf8eb0f3b4326ea24a1f8486d20eeae3758_2_1035x100.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/1/15028bf8eb0f3b4326ea24a1f8486d20eeae3758.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/15028bf8eb0f3b4326ea24a1f8486d20eeae3758_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-01-29 at 8.53.55.png</span><span class="informations">1276×124 9.97 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Thanks</p>

**Comment by ArturT** [comment]
<p>The 2 env variables that were introduced are still not unique :frowning: The number does not change if you re-run the workflow run.</p>
<div class="md-table">
<table>
<thead>
<tr>
<th>* GITHUB_RUN_ID</th>
<th>A unique number for each run within a repository. This number does not change if you re-run the workflow run.</th>
</tr>
</thead>
<tbody>
<tr>
<td>* GITHUB_RUN_NUMBER</td>
<td>A unique number for each run of a particular workflow in a repository. This number begins at 1 for the workflow’s first run, and increments with each new run. This number does not change if you re-run the workflow run.</td>
</tr>
</tbody>
</table>
</div><p>When someone reruns workflow I need to run a new split of tests between parallel jobs in knapsack pro tool on Github Actions. If I would use GITHUB_RUN_ID and someone reruns the workflow then my API would assume tests were already split for this GITHUB_RUN_ID. I would need GITHUB_UNIQUE_RUN_ID that would be always unique even for re-run workflow. This is how most CI providers generate CI build ID (it should be unique). <strong>Any chance you would introduce GITHUB_UNIQUE_RUN_ID ?</strong> Thank you.</p>

**Comment by liskin** [comment]
<p>I could use something like that as well: services such as coveralls accept a unique job id in their input and it would be good if they could link back to the specific job that generated the coverage.<br>
I’d suggest to name it <code>GITHUB_JOB_ID</code> and expose the job id that is already available through the API: <a href="https://developer.github.com/v3/actions/workflow-jobs/#get-a-job-for-a-workflow-run" rel="nofollow noopener">https://developer.github.com/v3/actions/workflow-jobs/#get-a-job-for-a-workflow-run</a></p>

**Comment by liskin** [comment]
<p>Hope I won’t jinx it, but it looks like this might finally get implemented: <a href="https://news.ycombinator.com/item?id=28472454" class="inline-onebox" rel="noopener nofollow ugc">No, there's no reason why that isn't available. This is a very sensible request ... | Hacker News</a> :slight_smile:</p>

**Reply by liskin** [reply]
I did jinx it apparently.

**Comment by joebowbeer** [comment]
<p>See <code>github.run_attempt</code> et al</p>
<p><a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context</a></p>

**Comment by RobSmyth** [comment]
The two things I consider as fundamental to a build system are:

* Reproducibility
* Traceability

To me a unique build number is a must have for traceability and therefore reproducibility. `github.run_attempt` and `github.run_number` do not provide a build number unless combined

In frustration I considering a build number of:

```
 <github.run_attempt>-<github.run_number>

to give something like ...

  1234-1, 12345-1, 12345-99
```

A build number for the whole repository (all workflows) would be fine. Just needs to:

* Be a single number (long standing convention)
* Increment on any build (including rebuilds)
* Uniquely identify a build

**Reply by aymather** [reply]
This seems so fundamental to the CI/CD build and deploy process it's wild to me that this is not a feature, but yes this is exactly what I am looking for as well.

**Comment by lkfortuna** [comment]
Hi there from Actions 👋  We do have a unique `id` property that is unique for every run of a workflow, even for reruns of that workflow. You can access this id via the REST api, you can see the id at the top of the example response here: https://docs.github.com/en/rest/actions/workflow-runs?apiVersion=2022-11-28#get-a-workflow-run

**Reply by ArturT** [reply]
@lkfortuna would it be possible to access the id via an env var?

**Reply by RobSmyth** [reply]
Hi @lkfortuna,

That `run id` is interesting. Do you know if it is number that is incremented on each build & each rebuild?

**Comment by RobSmyth** [comment]
Hi @lkfortuna,

Looking at that a bit more, I think the "unique ID" you found is `<github.run_number>`. I say this as it it used with the attempts count here:: https://docs.github.com/en/rest/actions/workflow-runs?apiVersion=2022-11-28#get-a-workflow-run-attempt.

If so, it does not increment on rebuilds and does not uniquely identify a build even though it may identify a GitHub "run" (which can be rebuilt).

**Reply by liskin** [reply]
Yeah, we've said it a million times here already - it's only the **workflow** run id that's exposed, and that's not unique enough. We need the **job** run id as well.

See e.g. this URL: https://github.com/liskin/strava-gear/actions/runs/10650245662/job/29521381751 - the first number is the workflow run id, the second is the job run id. The first one is available in the context and in an env var, the second isn't.

The fix is super simple - **just expose the second number somewhere, please.**

**Reply by RobSmyth** [reply]
Hi,

While only a workaround I'm using the `<github.run_attempt>` to get something like a built number. It is unique but it is so confusing for people who are so used to having a build number. Exposed as 2 identifiers (1234.1) in the semantic version so that it uses numerical precedence.

```
    - name: Build Test Project
      env:
          BUILD_NUMBER: ${{ github.run_number }}.${{ github.run_attempt }}
      run: |
          dotnet build ./TestApplication/TestApplication.sln \
                            --configuration $CONFIGURATION \
                            -p:BuildNumber=${{ env.BuildNumber }} --no-restore --verbosity minimal
```

Please GitHub ... provide a single build number.

**Comment by aeharding** [comment]
Recently ran into this problem because Google Play requires sequential build number for upload. So if a build #100 fails, next one (#101) succeeds, then I try to kick off #100 but it will always fail because it is #100, and rerun # doesn't change that.

Google play requires integer number too so there is really no workaround. Can't use `run_number.run_attempt`

https://github.com/aeharding/voyager/actions/runs/11876115553/attempts/2

**Reply by aeharding** [reply]
Basically all I need is an incrementing ID for any time any workflow is kicked off period. No repeatsies plz

**Reply by RobSmyth** [reply]
Hi @aeharding 

I'm not familiar with how Google Play handles build numbers. Can you give me a link to Google's documentation for what your doing?
It may help me with [versioning project I'm using](https://noetictools.github.io/Git2SemVer). I looked at the failure in the link you gave but it is not an area I'm familiar with.
